rfcs/20200422-concurrent-global-scratchpad-mode/README.md

```diff
@@ -0,0 +1,174 @@
+
+```cpp
+/// Scratchpad mode
+///
+concurrent = dnnl_scratchpad_mode_concurrent,
+///
+global = dnnl_scratchpad_mode_global,
+```
+
+## Compilation flag
+
+This proposal also aims to introduce a compilation flag
+`DNNL_DEFAULT_SCRATCHPAD_MODE`, which enables tweaking the scratchpad mode at
+compile-time for all primitives. It is proposed that this compilation flag will
+have the following three options:
+
+1. `DNNL_DEFAULT_SCRATCHPAD_MODE = DNNL_SCRATCHPAD_MODE_LIBRARY`. If specified,
+   the default scratchpad mode is equal to `scratchpad_mode::library`. This
+   maintain the current behavior of the library whereby each primitive decides
+   on its scratchpad mode.
+2. `DNNL_DEFAULT_SCRATCHPAD_MODE = DNNL_SCRATCHPAD_MODE_CONCURRENT`. If
+   specified, the default scratchpad mode is equal to
+   `scratchpad_mode::concurrent`, and enables a concurrent scratchpad for all
+   primitives.
+3. `DNNL_DEFAULT_SCRATCHPAD_MODE = DNNL_SCRATCHPAD_MODE_GLOBAL`. If specified,
+   the default scratchpad mode is equal to `scratchpad_mode::global`, and
+   enables a global scratchpad for all primitives.
+
+If `DNNL_DEFAULT_SCRATCHPAD_MODE` is not specified, the scratchpad mode in the
+library will default to `scratchpad_mode::library`.
+
+The above flags are expected to be beneficial in case a user does not require a
+per-primitive attribute and can rely on a global setting for scratchpad.
+Therefore, the users who do not use primitive attributes will benefit the most
+from the above compilation flags. However, the user who do use primitive
+attributes will not be impacted by usage of the above compilation flags since
+they only change the default `scratchpad_mode`, and it can be overriden by
+manually specifying a primitive attribute if necessary. 
+
+Given the above compilation flags, the proposal aims to replace
+`DNNL_ENABLE_CONCURRENT_EXEC` flag with `DNNL_DEFAULT_SCRATCHPAD_MODE =
+DNNL_SCRATCHPAD_MODE_CONCURRENT` since they are equivalent.
+
+## Usage models
+
+Since the new scratchpad modes are intended to complement existing one instead
+of replacing them, new usage models are assumed that may be favorable for the
+user in a particular instance. These usage models are briefly presented in the
+table below:
+
+| Scratchpad mode | Use case | Advantages |
+|-----------------|----------|------------|
+| **library**     | The library allocates scratchpad storage.<br>The mode is chosen by primitive implementation.                    | No API usage<br>     |
+| **concurrent**  | The library allocates scratchpad storage.<br>New scratchpad storage is created for each primitive.              | Thread-safety        |
+| **global**      | The library allocates scratchpad storage.<br>The storage is reused by primitives.                               | Reduced memory usage |
+| **user**        | The user will allocate scratchpad at primitive execution.<br>The scratchpad can be reused at user's discretion. | Full user control    |
+
```

Comment:
does it makes sense to clean up the naming so that they reflect better the allocation strategy? For example
`private`, `thread_local_pool` and `user` .

Reply:
I like the idea of replacing these names to something more intuitive. However in this case I think we will have to replace the type name in the library as well. Perhaps we can go with the following:

For the user:
- `concurrent` -> `primitive_private`
- `global` -> `thread_private`

Inside the library:
- `concurrent_scratchpad_t` -> `primitive_private_scratchpad_t`
- `global_scratchpad_t` -> `thread_private_scratchpad_t`

Reply:
Updated the RFC, please review.